code/datums/helper_datums/teleport.dm

```diff
@@ -97,7 +97,12 @@
 
 	var/turf/destturf
 	var/turf/curturf = get_turf(teleatom)
+	var/area/curarea = get_area(curturf)
 	var/area/destarea = get_area(destination)
+
+	if(!is_teleport_allowed(curturf.z) || curarea.tele_proof || !is_teleport_allowed(destturf.z) || destarea.tele_proof)

```

Comment:
For the record, though this is neat code-wise, and does address essentially every case... I am a little concerned it will lead to all sorts of horrible implications player-wise.

For example:

1. Prohibited teleports now simply fail. Unless the calling proc actually traps do_teleport and checks its return value, it won't know. Thus you'll get things like objects (hand teles, jaunters, etc) trying to create teleport portals from/to prohibited areas.... which works.... but then when you touch the portal, it doesn't teleport you. 

Personally I'd much prefer a solution that actually gave feedback in each case, along the lines of the teleport you are attempting being impossible.

But.... I guess this is the simple, universal, solution.